lib/dial/middleware.rb

```diff
@@ -48,13 +48,18 @@ def call env
       finish_time = Process.clock_gettime Process::CLOCK_MONOTONIC
       env[REQUEST_TIMING] = ((finish_time - start_time) * 1_000).round 2
 
+      csp_header = headers['Content-Security-Policy']
+      nonce = extract_nonce_from_csp(csp_header)
+
       body = String.new.tap do |str|
         rack_body.each { |chunk| str << chunk }
         rack_body.close if rack_body.respond_to? :close
-      end.sub "</body>", <<~HTML
-          #{Panel.html env, profile_out_filename, query_logs, ruby_vm_stat, gc_stat, gc_stat_heap, server_timing}
-        </body>
-      HTML
+
+        str.sub! "</body>", <<~HTML
```

Comment:
Hmm this is a good catch, I don't know why I'm copying the string here when I could mutate it directly. Probably some non-negligible perf gains just by doing that. Just have to keep in mind that this [could return `nil`](https://docs.ruby-lang.org/en/3.4/String.html#method-i-sub-21).

I'll action this as well separately.